File: utilities/shell/direnv-config/sdk/typescript/src/native.ts
import { readFile, writeFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';
import type { Backend } from './types.js';
import { getPath, setPath, deletePath } from './path.js';
import { layerPath } from './store.js';
import { resolveActive } from './resolve.js';
import { bumpVersion } from './version.js';

export class NativeBackend implements Backend {
  constructor(private storePath: string) {}

  async get(config: string, path?: string): Promise<unknown> {
    const activePath = join(this.storePath, config, '.active');
    const content = await readFile(activePath, 'utf-8');
    const root = parseYaml(content);

    if (!path) return root;
    return getPath(root, path);
  }

  async listConfigs(): Promise<string[]> {
    const metaPath = join(this.storePath, '.meta');
    const content = await readFile(metaPath, 'utf-8');
    const meta = parseYaml(content);
    return meta?.configs ?? [];
  }

  async set(config: string, key: string, value: string, layer = 'local', noBump = false): Promise<void> {
    const filePath = layerPath(this.storePath, config, layer);
    let doc: unknown = {};
    if (existsSync(filePath)) {
      const content = await readFile(filePath, 'utf-8');
      doc = parseYaml(content) ?? {};
    }

    let parsedValue: unknown;
    try {
      parsedValue = parseYaml(value);
      if (parsedValue === undefined) parsedValue = value;
    } catch {
      parsedValue = value;
    }

    doc = setPath(doc, key, parsedValue);
    await writeFile(filePath, stringifyYaml(doc), 'utf-8');
    await resolveActive(this.storePath, config);
    if (!noBump) {
      await bumpVersion(this.storePath);
    }
  }

  async unset(config: string, keys: string[], layer = 'local', noBump = false): Promise<void> {
    const filePath = layerPath(this.storePath, config, layer);
    if (!existsSync(filePath)) return;

    const content = await readFile(filePath, 'utf-8');
    const doc = parseYaml(content);
    if (doc === null || doc === undefined) return;

    for (const key of keys) {
      deletePath(doc, key);
    }

    await writeFile(filePath, stringifyYaml(doc), 'utf-8');
    await resolveActive(this.storePath, config);
    if (!noBump) {
      await bumpVersion(this.storePath);
    }
  }

  async bump(): Promise<number> {
    return bumpVersion(this.storePath);
  }
}

File: sdk/python/src/direnv_config/native.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml

from direnv_config.path import delete_path, get_path, set_path
from direnv_config.resolve import resolve_active
from direnv_config.store import ensure_config, layer_path
from direnv_config.version import bump_version


class NativeBackend:
    def __init__(self, store: Path) -> None:
        self._store = store

    def get(self, config: str, path: str | None = None) -> Any:
        active_file = self._store / config / ".active"
        if not active_file.exists():
            return None
        contents = active_file.read_text()
        data = yaml.safe_load(contents)
        if path is None:
            return data
        return get_path(data, path)

    def list_configs(self) -> list[str]:
        meta_file = self._store / ".meta"
        if not meta_file.exists():
            return []
        contents = meta_file.read_text()
        meta = yaml.safe_load(contents)
        if meta is None or not isinstance(meta, dict):
            return []
        configs = meta.get("configs", [])
        if not isinstance(configs, list):
            return []
        return [str(c) for c in configs]

    def set(
        self,
        config: str,
        key: str,
        value: str,
        layer: str = "local",
        no_bump: bool = False,
    ) -> None:
        ensure_config(self._store, config)
        lp = layer_path(self._store, config, layer)

        # Read existing layer or start fresh
        if lp.exists():
            doc = yaml.safe_load(lp.read_text())
            if doc is None:
                doc = {}
        else:
            doc = {}

        # Parse the value string
        parsed_value: Any
        if value is None:
            parsed_value = value
        else:
            try:
                parsed_value = yaml.safe_load(value)
            except yaml.YAMLError:
                parsed_value = value

        doc = set_path(doc, key, parsed_value)
        lp.write_text(yaml.dump(doc, default_flow_style=False))

        resolve_active(self._store, config)

        if not no_bump:
            bump_version(self._store)

    def unset(
        self,
        config: str,
        keys: list[str],
        layer: str = "local",
        no_bump: bool = False,
    ) -> None:
        lp = layer_path(self._store, config, layer)
        if not lp.exists():
            return

        doc = yaml.safe_load(lp.read_text())
        if doc is None:
            return

        for key in keys:
            delete_path(doc, key)

        lp.write_text(yaml.dump(doc, default_flow_style=False))

        resolve_active(self._store, config)

        if not no_bump:
            bump_version(self._store)

    def bump(self) -> int:
        return bump_version(self._store)
